core/shared/src/main/scala/zio/ZIOApp.scala

```diff
@@ -18,31 +18,37 @@ package zio
 import zio.internal._
 
 /**
- * The entry point for a ZIO application.
- *
- * {{{
- * import zio.ZIOApp
- * import zio.Console._
- *
- * object MyApp extends ZIOApp {
- *
- *   def run =
- *     for {
- *       _ <- printLine("Hello! What is your name?")
- *       n <- readLine
- *       _ <- printLine("Hello, " + n + ", good to meet you!")
- *     } yield ()
- * }
- * }}}
+ * An entry point for a ZIO application that allows sharing layers between
+ * applications. For a simpler version that uses the default ZIO environment
+ * see `ZIOAppDefault`.
  */
 trait ZIOApp { self =>
   @volatile private var shuttingDown = false
 
+  implicit def tag: Tag[Environment]
+
+  type Environment <: Has[_]
+
+  /**
+   * A layer that manages the acquisition and release of services necessary for
+   * the application to run.
+   */
+  def layer: ZLayer[ZEnv with Has[ZIOAppArgs], Any, Environment]
+
+  /**
+   * The main function of the application, which can access the command-line arguments through
+   * the `args` helper method of this class. If the provided effect fails for any reason, the
+   * cause will be logged, and the exit code of the application will be non-zero. Otherwise,
+   * the exit code of the application will be zero.
+   */
+  def run: ZIO[Environment with ZEnv with Has[ZIOAppArgs], Any, Any]
+
   /**
    * Composes this [[ZIOApp]] with another [[ZIOApp]], to yield an application that
    * executes the logic of both applications.
    */
-  final def <>(that: ZIOApp): ZIOApp = ZIOApp(self.run.zipPar(that.run), self.hook >>> that.hook)
+  final def <>(that: ZIOApp): ZIOApp =
+    ZIOApp(self.run.zipPar(that.run), self.layer +!+ that.layer, self.hook >>> that.hook)
```

Comment:
I know why the "unsafe" version is being used here `+|+`. It's probably ok in this case although it means the right hand side may clobber some implementations in the left, without the user being aware of this.

Reply:
Yes I think it is an unavoidable consequence of using the type member. Perhaps we could log a warning in the implementation of `Has` if we ever have implementations on the left or right and they are not the same object?